### PR TITLE
Scripts

### PR DESCRIPTION
**Pull request by eliascodes**
Related to TwinePlatform/data-sync#39

### Changes
* Add script to re-migrate user logs (and check for consistency after-the-fact)
* Add script to edit user roles
* Add script to clone db from dump-file (optional download from heroku)
* Change (most) scripts to Typescript using `ts-node`

<hr>

~~Only making this PR to expose the visit logs re-migration script (also the script to clone a db from a pg dump file).~~

~~The only file of interest in relation to TwinePlatform/data-sync#39 is ` bin/remigrate_visit_logs.ts` (see 0da330a).~~

~~I've run this script off a clone of the latest backup of the API database and the existing visitor app database, and am left with 3740 visit logs, compared with only 880 before (original visitor app has 3887, which sounds like close enough given the test accounts we've ignored; we can chase down the difference if we need to).~~

~~I figured you could work off this branch to run the scripts you need, without it needing to be merged, but if you'd prefer to get it merged, let me know and I can do a little tidying up!~~

~~PS: Only 2024dbb -> 0da330a were added in relation to TwinePlatform/data-sync#39, everything else was done a while ago and is only vaguely related. Happy to do some re-basing if you want to get only a subset merged.~~

### CODE REVIEW

**Comment by astroash** [comment]
@eliascodes why has my `bin/index.js` file been removed?

**Comment by eliascodes** [comment]
TS scripts can be run directly with `ts-node` (or with `npm run exc`).

e.g `ts-node ./bin/create_user.ts` or `npm run exc -- ./bin/create_user.ts --name=...`

File: bin/edit_user_role.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+/*
```
Comment (astroash) [inline bin/edit_user_role.ts]: Is this a script you have used on the production db?

File: bin/edit_user_role.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * script for editing an existing user
+ */
+import * as parse from 'minimist';
+import * as Knex from 'knex';
+import { Roles } from '../src/auth';
+import { getConfig } from '../config';
+
+process.on('unhandledRejection', (err) => { throw err; });
+
+const { userId, oid, add, rm } = parse(process.argv.slice(2));
```
Comment (astroash) [inline bin/edit_user_role.ts]: Can you add comments for how this script should be used. Eg what args are passed to it.

File: bin/remigrate_visit_logs.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+ */
+import * as parse from 'minimist';
+import * as Knex from 'knex';
+import { Visitors, Organisations } from '../src/models';
+import { getConfig } from '../config';
+
+
+process.on('unhandledRejection', (err) => { throw err; });
+
+const { _: [sourceUrl], ignore, ignoreNullAct } = parse(process.argv.slice(2));
```
Comment (astroash) [inline bin/remigrate_visit_logs.ts]: Does this work? You're using object destructuring on an array...

File: bin/remigrate_visit_logs.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        console.log(
+          `${visitor.id}@${org.name} :: `,
+          `Source logs: ${sourceLogsPreSync.length} | `,
+          `Target logs: ${trgtLogsPostSync.length}`,
+          `Inserting ${payload.length} logs`
+        );
+
+        try {
+          console.log(`Visitor ${visitor.qrCode} has ${trgtLogsPostSync.length} new logs`);
+          await trx('visit_log').del().where('user_account_id', visitor.id);
```
Comment (astroash) [inline bin/remigrate_visit_logs.ts]: What's going on here?

File: database/index.js [hunk 5]
```diff
@@ -97,14 +97,14 @@ exports.migrate = {
 };
 
 exports.add = {
-  apiToken: async (name, access, { env = process.env.NODE_ENV, client: _client } = {}) => {
+  apiToken: async (name, access, token, { env = process.env.NODE_ENV, client: _client } = {}) => {
```
Comment (astroash) [inline database/index.js]: This can probably be removed due to the decoupling of frontline.

File: bin/edit_user_role.ts (hunk 1 above)
Comment (eliascodes) [reply bin/edit_user_role.ts]: No. It started off as a script to edit user details but became too complicated, so I just changed it to manage roles. It's similar to #288 but takes arguments.

Comment (eliascodes) [reply bin/edit_user_role.ts]: Can also just delete though 🤷🏼‍♂️

File: bin/remigrate_visit_logs.ts (hunk 3 above)
Comment (eliascodes) [reply bin/remigrate_visit_logs.ts]: It works, the value of `_` is an array, so I'm using both array destructuring and object destructuring.

File: bin/remigrate_visit_logs.ts (hunk 4 above)
Comment (eliascodes) [reply bin/remigrate_visit_logs.ts]: Deleting all existing visit logs of the given visitor (re-added, along with any missing logs, on the next line).

File: bin/edit_user_role.ts (hunk 1 above)
Comment (astroash) [reply bin/edit_user_role.ts]: Up to you. Keep it if you think it’ll be useful in future.

Comment (eliascodes) [reply bin/edit_user_role.ts]: Lets keep it for now then

File: bin/edit_user_role.ts (hunk 2 above)
Comment (astroash) [reply bin/edit_user_role.ts]: @eliascodes can this be commented?

**Pull request merged**